### PR TITLE
cache on static input size/stride pr_0

### PR DESCRIPTION
**Pull request by jjsjann123**
While our kernels handle dynamic input sizes, we are now caching kernel selection and launch parameters on static sizes. This improves kernel launch latency for repeated input sizes.

This is now done at GraphCache level, where we record and encode every seen inputs

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp [hunk 1]
```diff
@@ -185,6 +186,36 @@ at::DimVector inversePermutation(
 
 } // namespace
 
+size_t InputsCodeLookup::getCode(const at::ArrayRef<IValue>& inputs) {
```
Comment (jjsjann123) [inline torch/csrc/jit/codegen/cuda/kernel_cache.cpp]: Please ignore the implementation here. This is a naive approach. I'm working on a trie to replace the stringstream.

**Comment by jjsjann123** [comment]
This one is ready for review.

File: torch/csrc/jit/codegen/cuda/executor.cpp [hunk 2]
```diff
@@ -70,7 +70,6 @@ void FusionExecutor::compileFusion(Fusion* fusion, CompileOptions options) {
       structured_code,
       (kernelNamespace() + "::" + kernelName()).c_str(),
       fusion_id_);
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/executor.cpp:72]: Check that `fusion_id_ > 0` ? (since we use this to tell if the kernel was compiled or not)

File: torch/csrc/jit/codegen/cuda/executor.h [hunk 3]
```diff
@@ -67,13 +85,14 @@ class TORCH_CUDA_API FusionExecutor : public NonCopyable {
       bool align_padding = false,
       uint64_t total = 0);
 
-  std::vector<at::Tensor> allocGlobalVals(EvaluationContext& ec);
+  // return a pair of vector of tensors, where tensors in the first vector are
+  // not initialized, while the second vector contains zero-initiliazed tensors
+  std::pair<std::vector<at::Tensor>, std::vector<at::Tensor>> allocGlobalVals(
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/executor.h]: please define a dedicated structure instead of returning std::pair<>

File: torch/csrc/jit/codegen/cuda/executor.cpp [hunk 4]
```diff
@@ -283,42 +283,106 @@ std::vector<at::Tensor> FusionExecutor::allocOutputs(EvaluationContext& ec) {
...
 
-  FusionGuard fg(&fusion_);
-
-  executor_utils::validateKernelInputs(&fusion_, inputs, options_.device);
+  ExecutorEntry* executor_entry = nullptr;
+  if (opt_code.has_value()) {
+    executor_entry = &executor_entry_lookup_[*opt_code];
+  }
 
+  FusionGuard fg(&fusion_);
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/executor.cpp:301]: why is this needed here?

File: torch/csrc/jit/codegen/cuda/executor.cpp [hunk 5]
```diff
@@ -283,42 +283,106 @@ std::vector<at::Tensor> FusionExecutor::allocOutputs(EvaluationContext& ec) {
...
+                                .device(options_.device);
+      empty_buffers.push_back(
+          at::empty(executor_entry->empty_buffer_sizes[i], tensor_options));
+    }
+    for (size_t i = 0; i < executor_entry->zero_buffer_sizes.size(); i++) {
+      auto tensor_options = at::TensorOptions()
+                                .dtype(executor_entry->zero_buffer_types[i])
+                                .device(options_.device);
+      zero_buffers.push_back(
+          at::zeros(executor_entry->zero_buffer_sizes[i], tensor_options));
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/executor.cpp:336]: not in the scope for this PR, but we really need to move away from using at::zeros() and at::empty() for buffer allocations: first, individual allocations have a significant overhead. second, and more important, at::zeros() may be just another operation in Python, but as far as I can tell it is horrific from native (C++) perspective: for each buffer, it will launch a kernel to zero-initialize the device memory!

@jjsjann123 , am I missing anything? if not, can you please open a Github issue to track this?

File: torch/csrc/jit/codegen/cuda/executor.cpp [hunk 6]
```diff
@@ -283,42 +283,106 @@ std::vector<at::Tensor> FusionExecutor::allocOutputs(EvaluationContext& ec) {
...
+    }
+    rand_offset = executor_entry->rand_offset;
+  } else {
+    executor_utils::validateKernelInputs(&fusion_, inputs, options_.device);
 
-  LaunchParams launch_params =
-      computeLaunchParams(inputs, launch_constraints, evaluation_context);
+    EvaluationContext evaluation_context =
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/executor.cpp:345]: can this be const?

File: torch/csrc/jit/codegen/cuda/executor.cpp [hunk 7]
```diff
@@ -283,42 +283,106 @@ std::vector<at::Tensor> FusionExecutor::allocOutputs(EvaluationContext& ec) {
...
+    if (outputs.empty() || outputs.size() != fusion_.outputs().size()) {
+      alloced_outputs = allocOutputs(evaluation_context);
+    }
+
+    executor_utils::validateKernelOutputs(
+        &fusion_, alloced_outputs, options_.device);
 
-  executor_utils::validateKernelOutputs(
-      &fusion_, alloced_outputs, options_.device);
+    std::tie(empty_buffers, zero_buffers) = allocGlobalVals(evaluation_context);
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/executor.cpp]: no need for tie if we use a dedicated structure

File: torch/csrc/jit/codegen/cuda/executor.cpp [hunk 8]
```diff
@@ -283,42 +283,106 @@ std::vector<at::Tensor> FusionExecutor::allocOutputs(EvaluationContext& ec) {
...
+
+    if (has_random_) {
+      rand_offset = 4 *
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/executor.cpp:367]: this non-trivial expression deserves a comment

File: torch/csrc/jit/codegen/cuda/executor.h [hunk 9]
```diff
@@ -28,18 +28,36 @@ class TORCH_CUDA_API FusionExecutor : public NonCopyable {
...
   // execute
   bool compiled() const {
-    return compiled_;
+    return fusion_id_ != -1;
+  };
+
+  // TODO: strides would also be important when we handle permutations in
+  //       codegen.
+  // struct used to hold necessary information to launch compiled kernel on a
+  // given input set;
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/executor.h]: nit: ; is not needed, right?

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp [hunk 10]
```diff
@@ -3,6 +3,7 @@
 #include <torch/csrc/jit/codegen/cuda/scheduler.h>
 #include <torch/csrc/jit/runtime/graph_executor.h>
 
+#include <nvToolsExt.h>
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_cache.cpp]: are you planning to check this in?

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp [hunk 11]
```diff
@@ -193,49 +224,57 @@ FusionExecutorCache::FusionExecutorCache(
...
-    Fusion fusion = *fusion_;
-    FusionGuard fg(&fusion);
-    TensorView* red_tv = nullptr;
-    for (auto expr : fusion.exprs()) {
-      if (expr->getExprType().has_value() &&
-          expr->getExprType().value() == ExprType::ReductionOp) {
-        red_tv = expr->outputs()[0]->as<TensorView>();
-        break;
+    const at::ArrayRef<IValue>& inputs,
+    const size_t code) {
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_cache.cpp]: 1. no const for parameters pls

2. what is the meaning of `code`? can we use a more descriptive name?

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp [hunk 12]
```diff
@@ -193,49 +224,57 @@ FusionExecutorCache::FusionExecutorCache(
...
+  if (code_to_fe_lookup_.count(code) == 0) {
+    // enter when we get a new input set. We need to search for compatible
+    // entries in cached `FusionExecutor` or compile new one as needed.
+
+    // caching strategy is different for pw-fusion and reduction-fusion.
+    if (has_reduction_) {
+      // copy the fusion, since each FusionExecutor needs to manipulate the
+      // fusion in order to generate kernel.
+      Fusion fusion = *fusion_;
+      FusionGuard fg(&fusion);
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_cache.cpp]: Is the guard needed here?

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp [hunk 13]
```diff
@@ -514,50 +553,65 @@ GraphCache::GraphCache(std::shared_ptr<Graph> graph)
...
+    appendFusionExecutorCache(
         InputsRequirement(graph_, toVector(reduction_axes_)));
   }
 }
 
 std::vector<at::Tensor> GraphCache::runGraphWithInputs(
     const at::ArrayRef<IValue>& inputs) {
-  InputsRequirement input_stack(inputs, toVector(reduction_axes_));
+  // get unique id `code` for given input set `inputs`;
+  size_t code = input_code_lookup_.getCode(inputs);
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_cache.cpp]: const. also maybe use a more descriptive name than `code`?

File: torch/csrc/jit/codegen/cuda/kernel_cache.h [hunk 14]
```diff
@@ -8,12 +8,27 @@
...
 namespace cuda {
 
+// Note, the uniqueness of the ide generated for a given input set is only local
+// to the instance of `InputsCodeLookup`.
+class InputsCodeLookup {
+ public:
+  // encode each unique input sets to an unique id;
+  size_t getCode(const at::ArrayRef<IValue>& inputs);
+
+ protected:
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_cache.h]: why protected and not private?

File: torch/csrc/jit/codegen/cuda/kernel_cache.h [hunk 15]
```diff
@@ -8,12 +8,27 @@
...
+  size_t current_id_ = 1;
+
+  // TODO: change this to a trie for efficiency;
+  std::unordered_map<std::string, size_t> inputs_code_lookup_;
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_cache.h]: just curious, did this show up in the execution profile as a significant overhead? I'd be surprised in a trie ends up significantly faster than a hash map

File: torch/csrc/jit/codegen/cuda/kernel_cache.h [hunk 16]
```diff
@@ -65,7 +80,8 @@ class FusionExecutorCache {
 
   // Execute fusion graph with given inputs, create `FusionExecutor` as needed;
   std::vector<at::Tensor> runFusionWithInputs(
-      const at::ArrayRef<IValue>& inputs);
+      const at::ArrayRef<IValue>& inputs,
+      const size_t code);
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/kernel_cache.h]: no const here

File: torch/csrc/jit/codegen/cuda/executor.cpp (hunk 5 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/executor.cpp:336]: at::empty & at::zeros go through the cuda caching allocator in PyTorch. So they maintain their own memory pools.

> for each buffer, it will launch a kernel to zero-initialize the device memory

I don't know how else that we can get 0-initialized memory other than a set memory, iiuc, which is a kernel by itself.

Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/executor.cpp:336]: We can combine all the allocations into a single cudaMalloc() + cudaMemset().

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/executor.cpp:336]: I think this is going to be a longer discussion. I'll open an issue to track it.

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/executor.cpp:336]: https://github.com/csarofeen/pytorch/issues/332

File: torch/csrc/jit/codegen/cuda/executor.cpp (hunk 4 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/executor.cpp:301]: hmmm, always nervous when I remove FusionGuard. But it does make sense to get rid of this one.
Note to myself what to search if segfault later.

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/executor.cpp:301]: It does indeed give me segfault. :/
I don't think we have a well defined / safe rule on when to set FusionGuard. I tried to dig out when it's needed and I think we used the fusion guard when query const scalar for extent.
But even after adding a fusion guard there and removing the one here, cpp tests are failing.

If we want to do a clean up of `FusionGuard` in integration, I think we need to at least have a guideline of when/where `FusionGuard` is needed at least.

File: torch/csrc/jit/codegen/cuda/executor.cpp (hunk 6 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/executor.cpp:345]: IIUC, it can't. We are passing evaluation_context as reference, which needs to be modified later when we do inference.

File: torch/csrc/jit/codegen/cuda/kernel_cache.h (hunk 15 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/kernel_cache.h]: I will profile it before doing anything specific.

Just to clarify, I'm not worried about hash lookup time vs trie traversal. It's actually the string concatenation that is needed to construct the key that I'm concerned.

**Comment by csarofeen** [comment]
This PR should also help with https://github.com/csarofeen/pytorch/issues/345

File: torch/csrc/jit/codegen/cuda/executor.cpp (hunk 4 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/executor.cpp:301]: Formalizing FusionGuard usage would be great, but fusion_ is Fusion Executors own copy of Fusion so not sure why we would care to get rid of this fusion guard. It makes sense generally that we will operate on it in this function.

File: torch/csrc/jit/codegen/cuda/executor.cpp (hunk 6 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/executor.cpp:345]: Might be able to make it const but I think we will replace this anyways with https://github.com/csarofeen/pytorch/pull/347

File: torch/csrc/jit/codegen/cuda/executor.cpp [hunk 17]
```diff
@@ -283,42 +285,111 @@ std::vector<at::Tensor> FusionExecutor::allocOutputs(EvaluationContext& ec) {
...
+  ExecutorEntry* executor_entry = nullptr;
+  if (opt_code.has_value()) {
+    executor_entry = &executor_entry_lookup_[*opt_code];
+  }
 
+  FusionGuard fg(&fusion_);
   c10::DeviceGuard dg(options_.device);
   auto stream = at::cuda::getCurrentCUDAStream();
 
-  EvaluationContext evaluation_context =
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/executor.cpp:297]: https://github.com/csarofeen/pytorch/pull/347
makes the evaluation of a fusion stateful. I'm wondering if we should cache the stateful evaluator here instead of everything. Maybe this introduces the chance of regressing though, and we should just move forward with the way you're already caching in this PR.

File: torch/csrc/jit/codegen/cuda/executor.cpp [hunk 18]
```diff
@@ -283,42 +285,111 @@ std::vector<at::Tensor> FusionExecutor::allocOutputs(EvaluationContext& ec) {
...
+      // works.
+      rand_offset = 4 *
+          (std::ceil(
+               alloced_outputs[0].numel() /
+               (4.0 * 128 * launch_params.gdimx())) + // NOLINT
+           1);
+    }
+
+    // record the the short-cut executor entry for the given input set;
+    if (executor_entry) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/executor.cpp:376]: Isn't executor_entry nullptr if we didn't have an entry? Shouldn't we be creating a new entry here if we didn't have one? Maybe I wasn't reading carefully enough.

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp [hunk 19]
```diff
@@ -193,49 +220,57 @@ FusionExecutorCache::FusionExecutorCache(
...
+    size_t unique_id) {
+  if (code_to_fe_lookup_.count(unique_id) == 0) {
+    // enter when we get a new input set. We need to search for compatible
+    // entries in cached `FusionExecutor` or compile new one as needed.
+
+    // caching strategy is different for pw-fusion and reduction-fusion.
+    if (has_reduction_) {
+      // copy the fusion, since each FusionExecutor needs to manipulate the
+      // fusion in order to generate kernel.
+      Fusion fusion = *fusion_;
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/kernel_cache.cpp:234]: We have to copy the fusion on every reduction run?!

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp [hunk 20]
```diff
@@ -514,50 +549,65 @@ GraphCache::GraphCache(std::shared_ptr<Graph> graph)
...
+      if (input_stack.complyWith(input_stacks_[i])) {
+        // found compliable fe_cache_ entry
+        fusion_executor_cache = fe_cache_[i].get();
+        // record short cut to designated fusion executor
+        code_to_index_lookup_[unique_id] = i;
+        break;
+      }
     }
+    if (!fusion_executor_cache) {
+      fusion_executor_cache = appendFusionExecutorCache(input_stack);
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/kernel_cache.cpp:582]: Does this mean you're creating a cache entry before we enter into runFusion?

File: torch/csrc/jit/codegen/cuda/executor.cpp (hunk 18 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/executor.cpp:376]: It looks like maybe you're always creating an entry at a higher level in the stack, assuming execution goes through JIT.

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/executor.cpp:376]: Are you resetting these values below even if we already have filled out values?

File: torch/csrc/jit/codegen/cuda/executor.cpp (hunk 17 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/executor.cpp:297]: I don't remember when we measure the number, if stateful evaluation still gives high latency. I though the number we have are pretty good.

File: torch/csrc/jit/codegen/cuda/executor.cpp (hunk 18 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/executor.cpp:376]: This one is quite confusing. It is to address our testing API where runFusion is used directly without an encoding for short-cut. At the beginning we have:
```
  ExecutorEntry* executor_entry = nullptr;
  if (opt_code.has_value()) {
    executor_entry = &executor_entry_lookup_[*opt_code];
  }
```

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/executor.cpp:376]: Basically this skips the short-cut recording when we do not provide a key to the short-cut map, in which case there's no entry provided.

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp (hunk 19 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/kernel_cache.cpp:234]: I think that's the case, since we did basically call `auto reduction_params = scheduleReduction(&fusion, inputs, red_tv);` to get reduction_params.
@kevinstephano in case I'm just using it wrong.

File: torch/csrc/jit/codegen/cuda/kernel_cache.cpp (hunk 20 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/kernel_cache.cpp:582]: The short answer is no.

This is the ugly bit, each level of cache has their own entry. At this point, we are creating an instance of `FusionExecutorCache` as well as a cache entry for `GraphCache` (the hosting cache).

But we are not creating any cache entry for nested structures. We only create cache entry below when we later call `fusion_executor_cache->runFusionWithInputs`

**Review by tlemo — APPROVED** [review]

File: torch/csrc/jit/codegen/cuda/executor.cpp (hunk 17 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/executor.cpp:297]: Stateful evaluation should be relatively low latency to reuse as it just holds the concrete values in an unordered map.

**Review by csarofeen — APPROVED** [review]
use empty_cuda

File: torch/csrc/jit/codegen/cuda/executor.cpp (hunk 18 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/executor.cpp:376]: Note to myself that it would be beneficial to have code path & stages in comments for better readability.
